execution/gethexec/wasmstorerebuilder.go

```diff
@@ -0,0 +1,132 @@
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/offchainlabs/nitro/arbos/arbosState"
+)
+
+var RebuildingPositionKey []byte = []byte("_rebuildingPosition")             // contains the codehash upto which rebuilding of wasm store was last completed. Initialized to common.Hash{} at the start
+var RebuildingStartBlockHashKey []byte = []byte("_rebuildingStartBlockHash") // contains the block hash of starting block when rebuilding of wasm store first began
+var RebuildingDone common.Hash = common.BytesToHash([]byte("_done"))         // indicates that the rebuilding is done, if RebuildingPositionKey holds this value it implies rebuilding was completed
+
+func ReadFromKeyValueStore[T any](store ethdb.KeyValueStore, key []byte) (T, error) {
+	var empty T
+	posBytes, err := store.Get(key)
+	if err != nil {
+		return empty, err
+	}
+	var val T
+	err = rlp.DecodeBytes(posBytes, &val)
+	if err != nil {
+		return empty, fmt.Errorf("error decoding value stored for key in the KeyValueStore: %w", err)
+	}
+	return val, nil
+}
+
+func WriteToKeyValueStore[T any](store ethdb.KeyValueStore, key []byte, val T) error {
+	valBytes, err := rlp.EncodeToBytes(val)
+	if err != nil {
+		return err
+	}
+	err = store.Put(key, valBytes)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// RebuildWasmStore function runs a loop looking at every codehash in diskDb, checking if its an activated stylus contract and
+// saving it to wasm store if it doesnt already exists. When errored it logs them and silently returns
+//
+// It stores the status of rebuilding to wasm store by updating the codehash (of the latest sucessfully checked contract) in
+// RebuildingPositionKey every 50 checks.
+//
+// It also stores a special value that is only set once when rebuilding commenced in RebuildingStartBlockHashKey as the block
+// time of the latest block when rebuilding was first called, this is used to avoid recomputing of assembly and module of
+// contracts that were created after rebuilding commenced since they would anyway already be added during sync.
+func RebuildWasmStore(ctx context.Context, wasmStore ethdb.KeyValueStore, l2Blockchain *core.BlockChain, position, rebuildingStartBlockHash common.Hash) error {
+	var err error
+	var stateDb *state.StateDB
+	tryLiveStateFallBackStateAt := func(header *types.Header) (*state.StateDB, error) {
+		if header == nil {
+			return nil, errors.New("trying to get state for a nil header")
+		}
+		if header.Root != (common.Hash{}) {
+			if err := l2Blockchain.TrieDB().Reference(header.Root, common.Hash{}); err != nil {
+				return l2Blockchain.StateAt(header.Root)
+			}
+		}
+		liveState, err := state.New(header.Root, l2Blockchain.StateCache(), nil)
+		if err != nil {
+			log.Info("Failed to get live state during rebuilding wasm store, falling back to StateAt", "err", err)
+			return l2Blockchain.StateAt(header.Root)
```

Comment:
having thought about it, I think it would be best to try using `stateAndHeaderFromHeader` to reuse the code and not to have to keep those places in sync

Reply:
I agree that exporting is the best approach here